File: 修订补充示例/第二轮/ch04_正文代码框.R
# 第 4 章 R：按章顺序提取教材正文代码框。
# 从“配套代码”根目录运行；data/processed 为冻结数据目录。
# 此文件不含审查断言；保留教材显示用的表达式。


# 教材：ch04_correlation_inference.tex:91；ch4_corr_r
library(readr); library(dplyr)
df <- read_csv("data/processed/chapter04_wdi_life_gdp_2024_economies.csv")
stopifnot(nrow(df) == 192, !anyDuplicated(df$country_code))

# Pearson 相关检验（默认 H0: rho = 0）
res <- cor.test(df$log_gdp_pc, df$life_expectancy,
                method = "pearson", conf.level = 0.95)
res$statistic   # t 统计量
res$p.value     # 参考模型下的 p 值
res$conf.int    # Fisher 变换 95% 置信区间

# 手工验证 Fisher 变换
r <- res$estimate; n <- nrow(df)
z  <- 0.5 * log((1 + r)/(1 - r))
se_z <- 1/sqrt(n - 3)
ci_z <- z + c(-1, 1) * qnorm(0.975) * se_z
ci_r <- (exp(2 * ci_z) - 1)/(exp(2 * ci_z) + 1)
round(ci_r, 3)   # 应与 res$conf.int 一致

# 按世界银行地区分类作描述性比较（并非大洲分类）
df %>% group_by(region) %>%
  summarise(r = cor(log_gdp_pc, life_expectancy), n = n())


# 教材：ch04_correlation_inference.tex:241；ch4_boot_r
set.seed(42)
B <- 9999
r_boot <- replicate(B, {
  idx <- sample(nrow(df), replace = TRUE)
  cor(df$log_gdp_pc[idx], df$life_expectancy[idx])
})
quantile(r_boot, c(0.025, 0.975))   # 百分位 Bootstrap 95% CI


# 教材：ch04_correlation_inference.tex:400；
x <- df$log_gdp_pc; y <- df$life_expectancy
z <- log(df$population)
ex <- resid(lm(x ~ z)); ey <- resid(lm(y ~ z))
partial <- cor(ex, ey)
semipartial <- cor(y, ex)
delta_r2 <- summary(lm(y ~ z + x))$r.squared -
            summary(lm(y ~ z))$r.squared
stopifnot(abs(semipartial^2 - delta_r2) < 1e-10)
c(partial = partial, semipartial = semipartial)


# 教材：ch04_correlation_inference.tex:479；
set.seed(42)
T_obs <- cor(df$log_gdp_pc, df$life_expectancy)
B <- 9999
T_perm <- replicate(B, cor(sample(df$log_gdp_pc),
                           df$life_expectancy))
p_perm <- (1 + sum(abs(T_perm) >= abs(T_obs))) / (B + 1)
p_perm  # 置换 p 值

File: 修订补充示例/第二轮/ch04_正文代码框.py
# 第4章Python：从教材按顺序提取；从仓库根目录运行。

# 代码框 1: book_chapters/ch04_correlation_inference.tex:118 / ch4_corr_py
import pandas as pd, numpy as np
from scipy import stats

df = pd.read_csv("data/processed/chapter04_wdi_life_gdp_2024_economies.csv")
assert len(df) == 192 and df["country_code"].is_unique

# Pearson 相关检验
r, pval = stats.pearsonr(df["log_gdp_pc"], df["life_expectancy"])
print(f"r = {r:.4f}, p = {pval:.2e}")

# 手工 Fisher 变换置信区间
n = len(df)
z  = np.arctanh(r)            # = 0.5 * log((1+r)/(1-r))
se = 1 / np.sqrt(n - 3)
ci_z = z + np.array([-1, 1]) * stats.norm.ppf(0.975) * se
ci_r = np.tanh(ci_z)          # 反变换回相关系数尺度
print(f"95% CI: [{ci_r[0]:.3f}, {ci_r[1]:.3f}]")

# 按地区分组
def corr_row(g):
    r2 = g["log_gdp_pc"].corr(g["life_expectancy"])
    return pd.Series({"r": r2, "n": len(g)})

df.groupby("region").apply(corr_row)


# 代码框 2: book_chapters/ch04_correlation_inference.tex:253 / ch4_boot_py
rng = np.random.default_rng(42)
r_boot = np.array([
    np.corrcoef(
        *df[["log_gdp_pc","life_expectancy"]]
          .sample(n=len(df), replace=True, random_state=rng).values.T
    )[0,1]
    for _ in range(9999)
])
np.percentile(r_boot, [2.5, 97.5])   # 百分位 Bootstrap CI


# 代码框 3: book_chapters/ch04_correlation_inference.tex:418 / ch04_correlation_inference_5
x = df["log_gdp_pc"].to_numpy()
y = df["life_expectancy"].to_numpy()
Z = np.column_stack([np.ones(len(df)), np.log(df["population"])])
ex = x - Z @ np.linalg.lstsq(Z, x, rcond=None)[0]
ey = y - Z @ np.linalg.lstsq(Z, y, rcond=None)[0]
partial = np.corrcoef(ex, ey)[0, 1]
semipartial = np.corrcoef(y, ex)[0, 1]
ZX = np.column_stack([Z, x])
e_full = y - ZX @ np.linalg.lstsq(ZX, y, rcond=None)[0]
delta_r2 = (ey @ ey - e_full @ e_full) / np.sum((y - y.mean())**2)
assert np.isclose(semipartial**2, delta_r2)
print(partial, semipartial)


# 代码框 4: book_chapters/ch04_correlation_inference.tex:495 / ch04_correlation_inference_7
import numpy as np
rng = np.random.default_rng(42)
x = df["log_gdp_pc"].to_numpy()
y = df["life_expectancy"].to_numpy()
B = 9999
t_obs = np.corrcoef(x, y)[0, 1]
t_perm = np.array([np.corrcoef(x, rng.permutation(y))[0, 1]
                   for _ in range(B)])
p_perm = (1 + np.sum(np.abs(t_perm) >= abs(t_obs))) / (B + 1)
print(p_perm)

